File: apps/ios/Barns/BarnsApp/Presentation/MyItems/ItemDetailViewModel.swift
import Combine

@MainActor
final class ItemDetailViewModel: ObservableObject {
    enum State: Equatable {
        case loading
        case loaded(ProductItem)
        case notFound
        case failed(String)
    }

    @Published private(set) var state: State = .loading
    /// Official read-only reference content (basic info + care guides) for the
    /// loaded item. `nil` until resolved; stays `nil` if nothing matches.
    @Published private(set) var officialContent: ItemOfficialContent?

    private let itemId: ProductItem.ID
    private let getProductItemDetailUseCase: GetProductItemDetailUseCase
    private let getGreeneryInfoUseCase: GetGreeneryInfoUseCase
    private let getCareGuidesUseCase: GetCareGuidesUseCase

    init(
        itemId: ProductItem.ID,
        getProductItemDetailUseCase: GetProductItemDetailUseCase,
        getGreeneryInfoUseCase: GetGreeneryInfoUseCase,
        getCareGuidesUseCase: GetCareGuidesUseCase
    ) {
        self.itemId = itemId
        self.getProductItemDetailUseCase = getProductItemDetailUseCase
        self.getGreeneryInfoUseCase = getGreeneryInfoUseCase
        self.getCareGuidesUseCase = getCareGuidesUseCase
    }

    func load() async {
        state = .loading
        officialContent = nil
        do {
            if let item = try await getProductItemDetailUseCase.execute(id: itemId) {
                state = .loaded(item)
                officialContent = await resolveOfficialContent(for: item)
            } else {
                state = .notFound
            }
        } catch {
            state = .failed("Unable to load this item. Please try again.")
        }
    }

    /// Resolves official basic information and care guides for the item.
    ///
    /// Linking is intentionally minimal: `ProductItem` has no `greeneryInfoId`
    /// yet, so a registered item is mapped to official content by its
    /// `categoryId` via `OfficialContentLink`. This is a documented temporary
    /// resolver — when `ProductItem` later carries an explicit official-content
    /// id, only this method changes. Official content is loaded through use
    /// cases (never direct repository/JSON access). Failures degrade to `nil`.
    private func resolveOfficialContent(for item: ProductItem) async -> ItemOfficialContent? {
        let link = OfficialContentLink.resolve(categoryId: item.categoryId)

        var overview: String?
        var lightPreference: String?
        var wateringOverview: String?
        if let infoId = link.greeneryInfoId,
           let info = try? await getGreeneryInfoUseCase.execute(id: infoId) {
            overview = info.overview
            lightPreference = info.lightPreference
            wateringOverview = info.wateringOverview
        }

        // Prefer the item's own care guides when present; otherwise fall back
        // to the category's default official guides.
        let guideIds = item.careGuideIds.isEmpty ? link.careGuideIds : item.careGuideIds
        let guides = (try? await getCareGuidesUseCase.execute(ids: guideIds)) ?? []
        let summaries = guides.map {
            ItemOfficialContent.CareGuideSummary(id: $0.id, title: $0.title, summary: $0.summary)
        }

        let content = ItemOfficialContent(
            overview: overview,
            lightPreference: lightPreference,
            wateringOverview: wateringOverview,
            careGuides: summaries
        )
        return content.isEmpty ? nil : content
    }
}

File: apps/android/app/src/main/java/com/barns/app/presentation/myitems/ItemDetailViewModel.kt
package com.barns.app.presentation.myitems

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.barns.app.domain.model.ProductItem
import com.barns.app.domain.usecase.myitems.ArchiveProductItemUseCase
import com.barns.app.domain.usecase.myitems.GetProductItemDetailUseCase
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

class ItemDetailViewModel(
    private val itemId: String,
    private val getProductItemDetailUseCase: GetProductItemDetailUseCase,
    private val officialContentResolver: ItemOfficialContentResolver,
    private val archiveProductItemUseCase: ArchiveProductItemUseCase,
) : ViewModel() {
    sealed interface State {
        object Loading : State
        data class Loaded(val item: ProductItem) : State
        object NotFound : State
        data class Failed(val message: String) : State
    }

    private val _state = MutableStateFlow<State>(State.Loading)
    val state: StateFlow<State> = _state.asStateFlow()

    /**
     * Official read-only reference content (basic info + care guides) for the
     * loaded item. `null` until resolved; stays `null` if nothing matches.
     */
    private val _officialContent = MutableStateFlow<ItemOfficialContent?>(null)
    val officialContent: StateFlow<ItemOfficialContent?> = _officialContent.asStateFlow()

    fun load() {
        _state.value = State.Loading
        _officialContent.value = null
        viewModelScope.launch {
            runCatching { getProductItemDetailUseCase.execute(itemId) }
                .onSuccess { item ->
                    if (item != null) {
                        _state.value = State.Loaded(item)
                        _officialContent.value = officialContentResolver.resolve(item)
                    } else {
                        _state.value = State.NotFound
                    }
                }
                .onFailure { _state.value = State.Failed("Unable to load this item. Please try again.") }
        }
    }

    /**
     * Archives the loaded greenery locally (soft action; no hard delete) and
     * invokes [onArchived] on success so the screen can return to the list.
     */
    fun archive(onArchived: () -> Unit) {
        val current = _state.value as? State.Loaded ?: return
        viewModelScope.launch {
            runCatching { archiveProductItemUseCase.execute(current.item) }
                .onSuccess { onArchived() }
                .onFailure {
                    _state.value = State.Failed("Unable to archive this item. Please try again.")
                }
        }
    }
}
